File: app/src/main/java/com/nfcalarmclock/statistics/NacStatisticsSettingFragment.java
package com.nfcalarmclock.statistics;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import com.google.android.material.button.MaterialButton;
import com.nfcalarmclock.R;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedPreferences;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Statistics fragment.
 * <p>
 * TODO: Highlighting the weekends in the dismiss/miss/snooze plot would be dope
 */
public class NacStatisticsSettingFragment
	extends Fragment
{

	/**
	 * Alarm statistics repository.
	 */
	private NacAlarmStatisticRepository mAlarmStatisticsRepository;

	/**
	 */
	public NacStatisticsSettingFragment()
	{
		super(R.layout.frg_statistics);

		this.mAlarmStatisticsRepository = null;
	}

	/**
	 */
	@Override
	public void onViewCreated(@NonNull View root, Bundle savedInstanceState)
	{
		super.onViewCreated(root, savedInstanceState);

		Context context = getContext();
		NacAlarmStatisticRepository repo = new NacAlarmStatisticRepository(context);
		this.mAlarmStatisticsRepository = repo;

		// Setup all the statistics
		this.setupDismissedAlarms(repo, root);
		this.setupSnoozedAlarms(repo, root);
		this.setupMissedAlarms(repo, root);
		this.setupCreatedAlarms(repo, root);
		this.setupDeletedAlarms(repo, root);
		this.setupCurrentAlarms(repo, root);
		this.setupStartedOnDate(repo, root);

		// Setup the reset button
		this.setupResetButton(root);

		// Setup all the views that need to use the theme color
		this.setupViewsWithThemeColor(root);
	}

	/**
	 * Get the alarm statistics repository.
	 *
	 * @return The alarm statistics repository.
	 */
	private NacAlarmStatisticRepository getAlarmStatisticsRepository()
	{
		return this.mAlarmStatisticsRepository;
	}

	/**
	 * Reset statistics.
	 */
	private void resetStatistics(View view)
	{
		View root = getView();
		NacAlarmStatisticRepository repo = this.getAlarmStatisticsRepository();

		// Delete all statistics
		repo.doDeleteAllCreated();
		repo.doDeleteAllDeleted();
		repo.doDeleteAllDismissed();
		repo.doDeleteAllMissed();
		repo.doDeleteAllSnoozed();

		// Change the text of when statistics started
		this.setupStartedOnDate(repo, root);
	}

	/**
	 * Setup the created alarm statistics.
	 */
	private void setupCreatedAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numCreated = repo.getCreatedCount();
		String text = String.valueOf(numCreated);

		TextView textview = root.findViewById(R.id.created_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the current alarm statistics.
	 */
	private void setupCurrentAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numCreated = repo.getCreatedCount();
		long numDeleted = repo.getDeletedCount();
		long numCurrent = numCreated - numDeleted;
		String text = String.valueOf(numCurrent);

		TextView textview = root.findViewById(R.id.current_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the deleted alarm statistics.
	 */
	private void setupDeletedAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numDeleted = repo.getDeletedCount();
		String text = String.valueOf(numDeleted);

		TextView textview = root.findViewById(R.id.deleted_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the dismissed alarm statistics.
	 */
	private void setupDismissedAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numDismissedTotal = repo.getDismissedCount();
		long numDismissedWithNfc = repo.getDismissedWithNfcCount();

		Locale locale = Locale.getDefault();
		String text = String.format(locale, "%1$s (%2$s NFC)", numDismissedTotal,
			numDismissedWithNfc);

		TextView textview = root.findViewById(R.id.dismissed_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the missed alarm statistics.
	 */
	private void setupMissedAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numMissed = repo.getMissedCount();
		String text = String.valueOf(numMissed);

		TextView textview = root.findViewById(R.id.missed_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the reset button.
	 */
	private void setupResetButton(View root)
	{
		MaterialButton resetButton = root.findViewById(R.id.reset_button);

		resetButton.setOnClickListener(this::resetStatistics);
	}

	/**
	 * Setup the snoozed alarm statistics.
	 */
	private void setupSnoozedAlarms(NacAlarmStatisticRepository repo, View root)
	{
		long numSnoozed = repo.getSnoozedCount();
		long snoozeDuration = repo.getSnoozedTotalDuration() / 60;

		Locale locale = Locale.getDefault();
		String text = String.format(locale, "%1$s (%2$s min)", numSnoozed,
			snoozeDuration);

		TextView textview = root.findViewById(R.id.snoozed_alarms_number);
		textview.setText(text);
	}

	/**
	 * Setup the date that statistics started on.
	 */
	private void setupStartedOnDate(NacAlarmStatisticRepository repo,
		View root)
	{
		Context context = getContext();
		NacSharedConstants cons = new NacSharedConstants(context);
		long timestamp = repo.getCreatedFirstTimestamp();
		String text;

		// Timestamp is a valid date
		if (timestamp > 0)
		{
			Locale locale = Locale.getDefault();
			String startedOnMessage = cons.getMessageStatisticsStartedOn();

			// Determine the format the date should be shown in
			Date dateStarted = new Date(timestamp);
			DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm z", locale);
			String dateText = dateFormat.format(dateStarted);

			// Set the text that will be shown in the textview
			text = String.format(locale, "%1$s %2$s", startedOnMessage, dateText);
		}
		// Empty or invalid timestamp, probably because statistics were reset
		else
		{
			// Set the text that will be shown in the textview
			text = "";
		}

		// Setup the textview that will show the date statistics were started
		// on, or that there were none found
		TextView textview = root.findViewById(R.id.statistics_started_on_date);

		textview.setText(text);
	}

	/**
	 * Setup views in the fragment so that they are using the correct theme color.
	 */
	private void setupViewsWithThemeColor(View root)
	{
		Context context = getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);

		// Get the views
		View divider1 = root.findViewById(R.id.divider1);
		View divider2 = root.findViewById(R.id.divider2);
		MaterialButton resetButton = root.findViewById(R.id.reset_button);

		// Get the theme color
		int themeColor = shared.getThemeColor();

		// Set the color of the dividers to the theme color
		divider1.setBackgroundColor(themeColor);
		divider2.setBackgroundColor(themeColor);

		// Set the color of the reset button to the theme color
		resetButton.setBackgroundColor(themeColor);
	}

}

File: app/src/main/java/com/nfcalarmclock/statistics/NacStatisticsSettingFragment.kt
package com.nfcalarmclock.statistics

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import com.google.android.material.button.MaterialButton
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Statistics fragment.
 * <p>
 * TODO: Highlighting the weekends in the dismiss/miss/snooze plot would be dope
 */
@AndroidEntryPoint
class NacStatisticsSettingFragment
	: Fragment(R.layout.frg_statistics)
{

	/**
	 * Statistic view model.
	 */
	private val statisticViewModel: NacAlarmStatisticViewModel by viewModels()

	/**
	 * Called when the view is created.
	 */
	override fun onViewCreated(root: View, savedInstanceState: Bundle?)
	{
		// Super
		super.onViewCreated(root, savedInstanceState)

		// Setup all the statistics
		lifecycleScope.launch {
			setupDismissedAlarms(root)
			setupSnoozedAlarms(root)
			setupMissedAlarms(root)
			setupCreatedAlarms(root)
			setupDeletedAlarms(root)
			setupCurrentAlarms(root)
			setupStartedOnDate(root)
		}

		// Setup the reset button
		setupResetButton(root)

		// Setup all the views that need to use the theme color
		setupViewsWithThemeColor(root)
	}

	/**
	 * Reset statistics.
	 */
	private fun resetStatistics()
	{
		// Delete all statistics
		statisticViewModel.deleteAllCreated()
		statisticViewModel.deleteAllDeleted()
		statisticViewModel.deleteAllDismissed()
		statisticViewModel.deleteAllMissed()
		statisticViewModel.deleteAllSnoozed()

		// Change the text of when statistics started
		lifecycleScope.launch {
			val root = requireView()

			setupStartedOnDate(root)
		}
	}

	/**
	 * Setup the created alarm statistics.
	 */
	private suspend fun setupCreatedAlarms(root: View)
	{
		// Get the text view
		val textview = root.findViewById<TextView>(R.id.created_alarms_number)

		// Determine how many alarms were created
		val numCreated = statisticViewModel.createdCount()

		// Set the text in the textview
		textview.text = numCreated.toString()
	}

	/**
	 * Setup the current alarm statistics.
	 */
	private suspend fun setupCurrentAlarms(root: View)
	{
		// Get the text view
		val textview = root.findViewById<TextView>(R.id.current_alarms_number)

		// Determine the current number of alarms
		val numCreated = statisticViewModel.createdCount()
		val numDeleted = statisticViewModel.deletedCount()
		val numCurrent = numCreated - numDeleted

		// Set the text in the textview
		textview.text = numCurrent.toString()
	}

	/**
	 * Setup the deleted alarm statistics.
	 */
	private suspend fun setupDeletedAlarms(root: View)
	{
		// Get the text view
		val textview = root.findViewById<TextView>(R.id.deleted_alarms_number)

		// Determine how many alarms were deleted
		val numDeleted = statisticViewModel.deletedCount()

		// Set the text in the textview
		textview.text = numDeleted.toString()
	}

	/**
	 * Setup the dismissed alarm statistics.
	 */
	private suspend fun setupDismissedAlarms(root: View)
	{
		// Get the text view
		val locale = Locale.getDefault()
		val textview = root.findViewById<TextView>(R.id.dismissed_alarms_number)

		// Determine how many alarms were dismissed and how many with NFC
		val numDismissedTotal = statisticViewModel.dismissedCount()
		val numDismissedWithNfc = statisticViewModel.dismissedWithNfcCount()

		// Determine the text to show in the textview
		val text = String.format(locale, "%1\$s (%2\$s NFC)", numDismissedTotal,
			numDismissedWithNfc)

		// Set the text in the textview
		textview.text = text
	}

	/**
	 * Setup the missed alarm statistics.
	 */
	private suspend fun setupMissedAlarms(root: View)
	{
		// Get the text view
		val textview = root.findViewById<TextView>(R.id.missed_alarms_number)

		// Determine how many alarms were missed
		val numMissed = statisticViewModel.missedCount()

		// Set the text in the textview
		textview.text = numMissed.toString()
	}

	/**
	 * Setup the reset button.
	 */
	private fun setupResetButton(root: View)
	{
		// Get the button
		val resetButton = root.findViewById<MaterialButton>(R.id.reset_button)

		// Set the listener
		// TODO: Show an "Are you sure?" dialog
		resetButton.setOnClickListener { resetStatistics() }
	}

	/**
	 * Setup the snoozed alarm statistics.
	 */
	private suspend fun setupSnoozedAlarms(root: View)
	{
		// Get the text view
		val locale = Locale.getDefault()
		val textview = root.findViewById<TextView>(R.id.snoozed_alarms_number)

		// Determine how many alarms were snoozed and how much time that was
		val numSnoozed = statisticViewModel.snoozedCount()
		val snoozeDuration = statisticViewModel.snoozedTotalDuration() / 60

		// Determine the text to show in the textview
		val text = String.format(locale, "%1\$s (%2\$s min)", numSnoozed,
			snoozeDuration)

		// Set the text in the textview
		textview.text = text
	}

	/**
	 * Setup the date that statistics started on.
	 */
	private suspend fun setupStartedOnDate(root: View)
	{
		// Get the created first timestamp
		val timestamp = statisticViewModel.createdFirstTimestamp()

		// Determine the text to show as the date the statistics started on
		val text: String =
			// Timestamp is a valid date
			if (timestamp > 0)
			{
				val locale = Locale.getDefault()
				val startedOnMessage = getString(R.string.message_statistics_started_on)

				// Determine the format the date should be shown in
				val dateStarted = Date(timestamp)
				val dateFormat: DateFormat =
					SimpleDateFormat("yyyy-MM-dd HH:mm z", locale)
				val dateText = dateFormat.format(dateStarted)

				// Set the text that will be shown in the textview
				String.format(locale, "%1\$s %2\$s", startedOnMessage, dateText)
			}
			else
			{
				// Set the text that will be shown in the textview
				""
			}

		// Setup the textview that will show the date statistics were started
		// on, or that there were none found
		val textview = root.findViewById<TextView>(R.id.statistics_started_on_date)

		textview.text = text
	}

	/**
	 * Setup views in the fragment so that they are using the correct theme color.
	 */
	private fun setupViewsWithThemeColor(root: View)
	{
		val shared = NacSharedPreferences(requireContext())

		// Get the views
		val divider1 = root.findViewById<View>(R.id.divider1)
		val divider2 = root.findViewById<View>(R.id.divider2)
		val resetButton = root.findViewById<MaterialButton>(R.id.reset_button)

		// Get the theme color
		val themeColor = shared.themeColor

		// Set the color of the dividers to the theme color
		divider1.setBackgroundColor(themeColor)
		divider2.setBackgroundColor(themeColor)

		// Set the color of the reset button to the theme color
		resetButton.setBackgroundColor(themeColor)
	}

}